rust/main/chains/hyperlane-sovereign/src/indexer.rs

```diff
@@ -0,0 +1,131 @@
+{
+    // These are the guys that need to be implemented by the concrete indexer
+    fn client(&self) -> &rest_client::SovereignRestClient;
+    fn decode_event(&self, event: &TxEvent) -> ChainResult<T>;
+    async fn sequence_at_slot(&self, slot: u32) -> ChainResult<Option<u32>>;
+    const EVENT_KEY: &'static str;
+
+    // Default implementation of Indexer<T>
+    async fn fetch_logs_in_range(
+        &self,
+        range: RangeInclusive<u32>,
+    ) -> ChainResult<Vec<(Indexed<T>, LogMeta)>> {
+        info!("Fetching {} logs in range: {:?}", Self::EVENT_KEY, range);
+
+        let mut results = Vec::new();
+
+        for batch_num in range {
+            let batch = self.client().get_batch(u64::from(batch_num)).await?;
+            let batch_hash = parse_hex_to_h256(&batch.hash, "invalid block hash")?;
+            for tx in &batch.txs {
+                let events = self.process_tx(tx, batch_hash)?;
+                results.extend(events);
+            }
+        }
+
+        Ok(results)
+    }
+
+    async fn get_finalized_block_number(&self) -> ChainResult<u32> {
+        info!(
+            "sov_indexer ({}): get_finalized_block_number",
+            Self::EVENT_KEY
+        );
+        let (_latest_slot, latest_batch) = self.client().get_latest_slot().await?;
+        Ok(latest_batch.unwrap_or_default())
+    }
+
+    async fn fetch_logs_by_tx_hash(
+        &self,
+        tx_hash: H512,
+    ) -> ChainResult<Vec<(Indexed<T>, LogMeta)>> {
+        let tx_hash: H256 = tx_hash.into();
+        let tx_hash = format!("0x{tx_hash:x}");
+        let tx = self.client().get_tx_by_hash(tx_hash).await?;
+        let batch = self.client().get_batch(tx.batch_number).await?;
+        let batch_hash = parse_hex_to_h256(&batch.hash, "invalid block hash")?;
+        self.process_tx(&tx, batch_hash)
+    }
+
+    // Default implementation of SequenceAwareIndexer<T>
+    async fn latest_sequence_count_and_tip(&self) -> ChainResult<(Option<u32>, u32)> {
+        let (latest_slot, latest_batch) = self.client().get_latest_slot().await?;
+        let sequence = self.sequence_at_slot(latest_slot).await?;
+
+        Ok((sequence, latest_batch.unwrap_or_default()))
+    }
+
+    // Helper function to process a single transaction
+    fn process_tx(&self, tx: &Tx, batch_hash: H256) -> ChainResult<Vec<(Indexed<T>, LogMeta)>> {
+        let mut results = Vec::new();
+
+        tx.events
+            .iter()
+            .filter(|e| e.key == Self::EVENT_KEY)
+            .try_for_each(|e| -> ChainResult<()> {
+                let (indexed_msg, meta) = self.process_event(tx, e, tx.batch_number, batch_hash)?;
+                info!(
+                    "Processed {} event : {:?} - Meta: {:?}",
+                    Self::EVENT_KEY,
+                    indexed_msg,
+                    meta
+                );
+                results.push((indexed_msg, meta));
+                Ok(())
+            })?;
+        Ok(results)
```

Comment:
So, would something like the below code work?
Removed the intermediate results vector and used collect() to create the result vector directly from the iterator. Would it be a bit faster to avoid pushing each element to a vector?

```suggestion
    tx.events
        .iter()
        .filter(|e| e.key == Self::EVENT_KEY)
        .map(|e| {
            let (indexed_msg, meta) = self.process_event(tx, e, tx.batch_number, batch_hash)?;
            info!(
                "Processed {} event: {:?} - Meta: {:?}",
                Self::EVENT_KEY,
                indexed_msg,
                meta
            );
            Ok((indexed_msg, meta))
        })
        .collect()
```